data_structures/rb_tree2.cpp

```diff
@@ -0,0 +1,1749 @@
+     *
+     * @param replacement_node the double-black node
+     */
+    void fix_double_black(RBTNode* replacement_node) {
+        // sibling not needed if replacement node is at the tree's root
+        if (replacement_node == root) {
+            replacement_node->color = COLOR_BLACK;
+        } else {
+            // now using sibling for below conditions
+            RBTNode* sibling = GetSibling(replacement_node);
+
+            // sibling is black - check sibling's kids
+            if (sibling->color == COLOR_BLACK) {
+                // at least one red child
+                if (has_red_child(sibling)) {
+                    // rotate/recolor to fix double-blackness
+                    CheckSiblingConditions(replacement_node, sibling);
+                } else {
+                    // no red children
+
+                    // shift blackness level up to parent
+                    sibling->color = COLOR_RED;
+                    replacement_node->color = COLOR_BLACK;
+                    replacement_node->parent->color++;
+                    // blackness shift may make parent double-black,
+                    // so fix if needed
+                    if (replacement_node->parent->color == COLOR_DOUBLE_BLACK) {
+                        fix_double_black(replacement_node->parent);
+                    }
+                }
+            } else if (sibling->color == COLOR_RED) {
+                // or replacement's sibling is red,
+                // rotate to give replacement a black sibling
+
+                // recoloring needed for rotation
+                sibling->parent->color = COLOR_RED;
+                sibling->color = COLOR_BLACK;
+
+                // rotate right for left sibling
+                if (ParentToMePtr(sibling) == sibling->parent->left) {
+                    RotateRight(sibling);
+                } else if (ParentToMePtr(sibling) == sibling->parent->right) {
+                    // rotate left for right sibling
+                    RotateLeft(sibling);
+                }
+                // try fixing double-blackness again after
+                // giving replacement node a black sibling
+                fix_double_black(replacement_node);
+            }
+        }
+    }
+
+    /**
+     * @brief Checks if passed node has a red child.
+     *
+     * @param n
+     * @return true
+     * @return false
+     */
+    bool has_red_child(RBTNode* n) {
+        RBTNode* n_left = n->left;
+        RBTNode* n_right = n->right;
+
+        // at least one red child
+        return (n_left != nullptr && n_left->color == COLOR_RED) ||
+               (n_right != nullptr && n_right->color == COLOR_RED);
+    }
+
+    /**
+     * @brief Rotates/recolors tree based on four sibling cases
+     * (left left, left right, right right, right left) to
+     * fix double-blackness issue in tree.
+     *
+     * *Note: Assumed that the parameters given will be correct for this
```

Comment:
```suggestion
     * @note Assumed that the parameters given will be correct for this
```
doxygen has tags for this!